### ISSUE TITLE
Inference with the trained model

### ISSUE THREAD

**Issue by igormis** [billpku/NLP_In_Action#5]
I have trained the model and saved the model, I have the following three files:
- config.json
- pytorch_model.bin
- vocab.txt
Now I would like to make inference on new samples, I have problem to load the model, any info on this?

**Comment by billpku** [comment]
First of all, If you can post the error message, then I can help you with more info.

Now I suppose that you have already trained a BERT NER model successfully.
The NER model can identify 20 tags.
To load your trained model, you need 2 files:

- config.json
- pytorch_model.bin

You need to put these files into the same folder, like: models/NER
After doing this:
you can load the mode with this code:

```pyhton
from pytorch_transformers import BertForTokenClassification

# if you use transformers
# from transformers import BertForTokenClassification 

model = BertForTokenClassification.from_pretrained('models/NER',num_labels=20)
```

Hope this can help you.

**Comment by igormis** [comment]
Tnx. I do not have any errors. I can load the model like this. But. if I have sentence (text) as an input how do I send to the model and obtain the NER output?

**Comment by billpku** [comment]
> Tnx. I do not have any errors. I can load the model like this. But. if I have sentence (text) as an input how do I send to the model and obtain the NER output?

Sorry for the late reply.

It seem like you must make your inputs into embeddings first, the inference process is like this:

1. load model and tokenizer

2. Tokenize your word input into token id embedding. E.g: input = "[CLS] I have a dream. [SEP]", after tokenize into id embedding, the result may be like this "[101,3,4,5,7,32,103]" 

3. Set input attention_masks_embedding.(with token mask is 1, pad token is 0)

4. Get the model predict result:
```python
# model is the model you load before
# input_ids is your input token id embedding
# attention_mask is your input attention masks embedding
with torch.no_grad():
        outputs = model(input_ids, token_type_ids=None,
        attention_mask=input_mask,)
        # For eval mode, the first result of outputs is logits
        logits = outputs[0] 
    
    # Get NER predict result
    logits = torch.argmax(F.log_softmax(logits,dim=2),dim=2)
    logits = logits.detach().cpu().numpy()
    
    
    # Get NER true result
    label_ids = label_ids.to('cpu').numpy()
    
    
    # Only predict the real word, mark=0, will not calculate
    input_mask = input_mask.to('cpu').numpy()
```

You can check also my notebook "Make training data" and "Eval model" parts, they cover how to make input into embeddings and how to  get the model predict result.

And your issue also remind me that I need to include the inference process in my notebook as a whole "NLP in action" process.

Thank you for your suggestion.

**Comment by marcusau** [comment]
based on your code mentioned above, that means , i can see the NER result in "label_ids "?????

Thanks a lot

**Comment by billpku** [comment]
> I have trained the model and saved the model, I have the following three files:
> 
> * config.json
> * pytorch_model.bin
> * vocab.txt
>   Now I would like to make inference on new samples, I have problem to load the model, any info on this?

@igormis 

Hi,

 I had updated my notebook adding inference support:
- [NER_with_BERT.ipynb](https://github.com/billpku/NLP_In_Action/blob/master/NER_with_BERT.ipynb)

Feel free to leave a comment.

**Comment by igormis** [comment]
Tnx @billpku will check the notebook
